lsp/nls/src/requests/hover.rs

```diff
@@ -1,68 +1,169 @@
-use log::debug;
 use lsp_server::{RequestId, Response, ResponseError};
 use lsp_types::{Hover, HoverContents, HoverParams, LanguageString, MarkedString, Range};
-use nickel_lang_core::position::TermPos;
+use nickel_lang_core::{
+    identifier::Ident,
+    position::RawSpan,
+    term::{record::FieldMetadata, LabeledType, RichTerm, Term, UnaryOp},
+    typ::Type,
+};
 use serde_json::Value;
 
 use crate::{
     cache::CacheExt,
     diagnostic::LocationCompat,
+    field_walker::{FieldHaver, FieldResolver},
+    identifier::LocIdent,
     server::Server,
-    trace::{Enrich, Trace},
 };
 
+#[derive(Debug)]
+struct HoverData {
+    values: Vec<RichTerm>,
+    metadata: Vec<FieldMetadata>,
+    span: RawSpan,
+    ty: Option<Type>,
+}
+
+fn annotated_contracts(rt: &RichTerm) -> &[LabeledType] {
+    match rt.as_ref() {
+        Term::Annotated(annot, _) => &annot.contracts,
+        _ => &[],
+    }
+}
+
+fn nickel_string(s: String) -> MarkedString {
+    MarkedString::LanguageString(LanguageString {
+        language: "nickel".to_owned(),
+        value: s,
+    })
+}
+
+fn values_and_metadata_from_field(
+    parents: Vec<FieldHaver>,
+    ident: Ident,
+) -> (Vec<RichTerm>, Vec<FieldMetadata>) {
+    let mut values = Vec::new();
+    let mut metadata = Vec::new();
+    for parent in parents {
+        if let FieldHaver::RecordTerm(r) = parent {
+            if let Some(field) = r.fields.get(&ident) {
+                values.extend(field.value.iter().cloned());
+                metadata.push(field.metadata.clone());
+            }
+        }
+    }
+    (values, metadata)
+}
+
+fn ident_hover(ident: LocIdent, server: &Server) -> Option<HoverData> {
+    let ty = server.lin_registry.get_type_for_ident(&ident).cloned();
+    let span = ident.pos.into_opt()?;
+    let mut ret = HoverData {
+        values: Vec::new(),
+        metadata: Vec::new(),
+        span,
+        ty,
+    };
+
+    if let Some(def) = server.lin_registry.get_def(&ident) {
+        let resolver = FieldResolver::new(server);
+        if let Some(((last, path), val)) = def.path.split_last().zip(def.value.as_ref()) {
+            let parents = resolver.resolve_term_path(val, path);
+            let (values, metadata) = values_and_metadata_from_field(parents, *last);
+            ret.values = values;
+            ret.metadata = metadata;
+        } else if def.path.is_empty() {
+            ret.values.extend(def.value.iter().cloned());
+        }
+    }
+
+    Some(ret)
+}
+
+fn term_hover(rt: &RichTerm, server: &Server) -> Option<HoverData> {
```

Comment:
It's a detail, and probably not for this PR, but I wonder if all those `term_xxx`, `type_xxx`, `contract_xxx` for several `xxx` (finding completion information, getting hover information, and so on) wouldn't be a bit more organized as traits. For example a `trait Hover { .. hover(..); }` that is implemented by `RichTerm`, `Ident` and whatnot.